File: app/src/main/java/com/quran/labs/androidquran/ui/helpers/QuranListAdapter.java
package com.quran.labs.androidquran.ui.helpers;

import android.content.Context;
import android.graphics.PorterDuff;
import android.support.v7.widget.RecyclerView;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.dao.Bookmark;
import com.quran.labs.androidquran.dao.Tag;
import com.quran.labs.androidquran.ui.QuranActivity;
import com.quran.labs.androidquran.util.QuranUtils;
import com.quran.labs.androidquran.widgets.JuzView;
import com.quran.labs.androidquran.widgets.TagsViewGroup;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class QuranListAdapter extends
    RecyclerView.Adapter<QuranListAdapter.HeaderHolder>
    implements View.OnClickListener, View.OnLongClickListener {

  private Context context;
  private LayoutInflater inflater;
  private QuranRow[] elements;
  private RecyclerView recyclerView;
  private SparseBooleanArray checkedState;
  private QuranTouchListener touchListener;
  private Map<Long, Tag> tagMap;
  private boolean showTags;
  private boolean isEditable;

  public QuranListAdapter(Context context, RecyclerView recyclerView,
                          QuranRow[] items, boolean isEditable) {
    inflater = LayoutInflater.from(context);
    this.recyclerView = recyclerView;
    elements = items;
    this.context = context;
    checkedState = new SparseBooleanArray();
    this.isEditable = isEditable;
  }

  public long getItemId(int position) {
    return position;
  }

  @Override
  public int getItemCount() {
    return elements.length;
  }

  private QuranRow getQuranRow(int position) {
    return elements[position];
  }

  public boolean isItemChecked(int position) {
    return checkedState.get(position);
  }

  public void setItemChecked(int position, boolean checked) {
    checkedState.put(position, checked);
    notifyItemChanged(position);
  }

  public List<QuranRow> getCheckedItems() {
    final List<QuranRow> result = new ArrayList<>();
    final int count = checkedState.size();
    final int elements = getItemCount();
    for (int i = 0; i < count; i++) {
      final int key = checkedState.keyAt(i);
      // TODO: figure out why sometimes elements > key
      if (checkedState.get(key) && elements > key) {
        result.add(getQuranRow(key));
      }
    }
    return result;
  }

  public void uncheckAll() {
    checkedState.clear();
    notifyDataSetChanged();
  }

  public void setElements(QuranRow[] elements, Map<Long, Tag> tagMap) {
    this.elements = elements;
    this.tagMap = tagMap;
  }

  public void setShowTags(boolean showTags) {
    this.showTags = showTags;
  }

  private void bindRow(HeaderHolder vh, int position) {
    ViewHolder holder = (ViewHolder) vh;

    final QuranRow item = elements[position];
    bindHeader(vh, position);
    holder.number.setText(
        QuranUtils.getLocalizedNumber(context, item.sura));

    holder.metadata.setVisibility(View.VISIBLE);
    holder.metadata.setText(item.metadata);
    holder.tags.setVisibility(View.GONE);

    if (item.juzType != null) {
      holder.image.setImageDrawable(
          new JuzView(context, item.juzType, item.juzOverlayText));
      holder.image.setVisibility(View.VISIBLE);
      holder.number.setVisibility(View.GONE);
    } else if (item.imageResource == null) {
      holder.number.setVisibility(View.VISIBLE);
      holder.image.setVisibility(View.GONE);
    } else {
      holder.image.setImageResource(item.imageResource);
      if (item.imageFilterColor == null) {
        holder.image.setColorFilter(null);
      } else {
        holder.image.setColorFilter(
            item.imageFilterColor, PorterDuff.Mode.SRC_ATOP);
      }
      holder.image.setVisibility(View.VISIBLE);
      holder.number.setVisibility(View.GONE);

      List<Tag> tags = new ArrayList<>();
      Bookmark bookmark = item.bookmark;
      if (bookmark != null && !bookmark.tags.isEmpty() && showTags) {
        for (int i = 0, bookmarkTags = bookmark.tags.size(); i < bookmarkTags; i++) {
          Long tagId = bookmark.tags.get(i);
          Tag tag = tagMap.get(tagId);
          if (tag != null) {
            tags.add(tag);
          }
        }
      }

      if (tags.isEmpty()) {
        holder.tags.setVisibility(View.GONE);
      } else {
        holder.tags.setTags(tags);
        holder.tags.setVisibility(View.VISIBLE);
      }
    }
  }

  private void bindHeader(HeaderHolder holder, int pos) {
    final QuranRow item = elements[pos];
    holder.title.setText(item.text);
    if (item.page == 0) {
      holder.pageNumber.setVisibility(View.GONE);
    } else {
      holder.pageNumber.setVisibility(View.VISIBLE);
      holder.pageNumber.setText(
          QuranUtils.getLocalizedNumber(context, item.page));
    }
    holder.setChecked(isItemChecked(pos));

    final boolean enabled = isEnabled(pos);
    holder.setEnabled(enabled);
  }

  @Override
  public HeaderHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    if (viewType == 0) {
      final View view = inflater.inflate(R.layout.index_header_row, parent, false);
      return new HeaderHolder(view);
    } else {
      final View view = inflater.inflate(R.layout.index_sura_row, parent, false);
      return new ViewHolder(view);
    }
  }

  @Override
  public void onBindViewHolder(HeaderHolder viewHolder, int position) {
    final int type = getItemViewType(position);
    if (type == 0) {
      bindHeader(viewHolder, position);
    } else {
      bindRow(viewHolder, position);
    }
  }

  @Override
  public int getItemViewType(int position) {
    return elements[position].isHeader() ? 0 : 1;
  }

  private boolean isEnabled(int position) {
    final QuranRow selected = elements[position];
    return !isEditable ||                     // anything in surahs or juzs
        selected.isBookmark() ||              // actual bookmarks
        selected.rowType == QuranRow.NONE ||  // the actual "current page"
        selected.isBookmarkHeader();          // tags
  }

  public void setQuranTouchListener(QuranTouchListener listener) {
    touchListener = listener;
  }

  @Override
  public void onClick(View v) {
    final int position = recyclerView.getChildAdapterPosition(v);
    if (position != RecyclerView.NO_POSITION) {
      final QuranRow element = elements[position];
      if (touchListener == null) {
        ((QuranActivity) context).jumpTo(element.page);
      } else {
        touchListener.onClick(element, position);
      }
    }
  }

  @Override
  public boolean onLongClick(View v) {
    if (touchListener != null) {
      final int position = recyclerView.getChildAdapterPosition(v);
      if (position != RecyclerView.NO_POSITION) {
        return touchListener.onLongClick(elements[position], position);
      }
    }
    return false;
  }

  class HeaderHolder extends RecyclerView.ViewHolder {

    TextView title;
    TextView pageNumber;
    View view;

    HeaderHolder(View itemView) {
      super(itemView);
      view = itemView;
      title = (TextView) itemView.findViewById(R.id.title);
      pageNumber = (TextView) itemView.findViewById(R.id.pageNumber);
    }

    void setEnabled(boolean enabled) {
      view.setEnabled(enabled);
      itemView.setOnClickListener(enabled ? QuranListAdapter.this : null);
      itemView.setOnLongClickListener(isEditable && enabled ? QuranListAdapter.this : null);
    }

    void setChecked(boolean checked) {
      view.setActivated(checked);
    }
  }

  private class ViewHolder extends HeaderHolder {

    TextView number;
    TextView metadata;
    ImageView image;
    TagsViewGroup tags;

    ViewHolder(View itemView) {
      super(itemView);
      metadata = (TextView) itemView.findViewById(R.id.metadata);
      number = (TextView) itemView.findViewById(R.id.suraNumber);
      image = (ImageView) itemView.findViewById(R.id.rowIcon);
      tags = (TagsViewGroup) itemView.findViewById(R.id.tags);
    }
  }

  public interface QuranTouchListener {

    void onClick(QuranRow row, int position);

    boolean onLongClick(QuranRow row, int position);
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/helpers/QuranListAdapter.kt
package com.quran.labs.androidquran.ui.helpers

import android.content.Context
import android.graphics.PorterDuff
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.core.util.set
import androidx.recyclerview.widget.RecyclerView
import com.quran.data.model.bookmark.Tag
import com.quran.labs.androidquran.R
import com.quran.labs.androidquran.common.ui.core.CollectionNames
import com.quran.labs.androidquran.ui.QuranActivity
import com.quran.labs.androidquran.util.QuranUtils
import com.quran.labs.androidquran.view.JuzView
import com.quran.labs.androidquran.view.TagsViewGroup
import java.text.SimpleDateFormat
import java.util.Date
import androidx.core.util.size

class QuranListAdapter(
  private val context: Context,
  private val recyclerView: RecyclerView,
  private var elements: Array<QuranRow>,
  private val isEditable: Boolean
) : RecyclerView.Adapter<QuranListAdapter.HeaderHolder>(),
  View.OnClickListener, View.OnLongClickListener {

  private val inflater = LayoutInflater.from(context)
  private val checkedState = SparseBooleanArray()
  private val locale = QuranUtils.getCurrentLocale()
  private val emptyColorAlpha =
    ResourcesCompat.getFloat(context.resources, R.dimen.empty_highlight_alpha)
  private var tagMap: Map<String, Tag> = emptyMap()
  private var showTags = false
  private var showDate = false

  private var touchListener: QuranTouchListener? = null

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HeaderHolder {
    return when (viewType) {
      VIEW_TYPE_HEADER ->
        HeaderHolder(inflater.inflate(R.layout.index_header_row, parent, false))

      VIEW_TYPE_COLLECTION_HEADER ->
        CollectionHeaderHolder(
          inflater.inflate(R.layout.index_collection_header_row, parent, false)
        )

      VIEW_TYPE_HIGHLIGHT_COLOR ->
        HighlightColorHolder(
          inflater.inflate(R.layout.index_highlight_color_row, parent, false)
        )

      else -> ViewHolder(inflater.inflate(R.layout.index_sura_row, parent, false))
    }
  }

  override fun onBindViewHolder(holder: HeaderHolder, position: Int) {
    when (getItemViewType(position)) {
      VIEW_TYPE_HEADER -> bindHeader(holder, position)
      VIEW_TYPE_COLLECTION_HEADER -> bindCollectionHeader(holder as CollectionHeaderHolder, position)
      VIEW_TYPE_HIGHLIGHT_COLOR -> bindHighlightColor(holder as HighlightColorHolder, position)
      else -> bindRow(holder, position)
    }
  }

  override fun getItemCount(): Int = elements.size

  override fun getItemId(position: Int): Long = position.toLong()

  override fun getItemViewType(position: Int): Int {
    val element = elements[position]
    return when {
      element.isBookmarkHeader -> VIEW_TYPE_COLLECTION_HEADER
      element.isHeader -> VIEW_TYPE_HEADER
      element.isHighlightColor -> VIEW_TYPE_HIGHLIGHT_COLOR
      else -> VIEW_TYPE_ROW
    }
  }

  override fun onClick(v: View) {
    val position = recyclerView.getChildAdapterPosition(v)
    if (position != RecyclerView.NO_POSITION) {
      val element = elements[position]
      if (touchListener == null) {
        (context as QuranActivity).jumpTo(element.page)
      } else {
        touchListener?.onClick(element, position)
      }
    }
  }

  override fun onLongClick(v: View): Boolean {
    touchListener?.let { listener ->
      val position = recyclerView.getChildAdapterPosition(v)
      if (position != RecyclerView.NO_POSITION) {
        return listener.onLongClick(elements[position], position)
      }
    }
    return false
  }

  fun setElements(elements: Array<QuranRow>) {
    this.elements = elements
    notifyDataSetChanged()
  }

  fun isItemChecked(position: Int): Boolean = checkedState[position]

  fun setItemChecked(position: Int, checked: Boolean) {
    checkedState[position] = checked
    notifyItemChanged(position)
  }

  fun uncheckAll() {
    checkedState.clear()
    notifyDataSetChanged()
  }

  fun getCheckedItems(): List<QuranRow> {
    val result = ArrayList<QuranRow>()
    val count = checkedState.size
    val elements = itemCount
    for (i in 0 until count) {
      val key = checkedState.keyAt(i)
      // TODO: figure out why sometimes elements > key
      if (checkedState[key] && elements > key) {
        result.add(getQuranRow(key))
      }
    }
    return result
  }

  fun setQuranTouchListener(listener: QuranTouchListener) {
    touchListener = listener
  }

  fun setElements(elements: Array<QuranRow>, tagMap: Map<String, Tag>) {
    this.elements = elements
    this.tagMap = tagMap
  }

  fun setShowTags(showTags: Boolean) {
    this.showTags = showTags
  }

  fun setShowDate(showDate: Boolean) {
    this.showDate = showDate
  }

  private fun getQuranRow(position: Int): QuranRow = elements[position]

  private fun bindRow(vh: HeaderHolder, position: Int) {
    val holder = vh as ViewHolder
    bindHeader(vh, position)
    val item = elements[position]

    with(holder) {
      number.text = QuranUtils.getLocalizedNumber(item.sura)
      metadata.visibility = View.VISIBLE
      metadata.text = item.metadata
      tags.visibility = View.GONE


      when {
        item.juzType != null -> {
          image.setImageDrawable(
            JuzView(context, item.juzType, item.juzOverlayText)
          )
          image.visibility = View.VISIBLE
          number.visibility = View.GONE
        }
        item.imageResource == null -> {
          number.visibility = View.VISIBLE
          image.visibility = View.GONE
        }
        else -> {
          image.setImageResource(item.imageResource)
          if (item.imageFilterColorResource == null) {
            image.colorFilter = null
          } else {
            image.setColorFilter(
              ContextCompat.getColor(context, item.imageFilterColorResource), PorterDuff.Mode.SRC_ATOP
            )
          }

          if (showDate) {
            val date = SimpleDateFormat("MMM dd, HH:mm", locale)
              .format(Date(item.dateAddedInMillis))
            holder.metadata.text = buildString {
              append(item.metadata)
              append(" - ")
              append(date)
            }
          }

          image.visibility = View.VISIBLE
          number.visibility = View.GONE

          val tagList = ArrayList<Tag>()
          val bookmark = item.bookmark
          if (bookmark != null && bookmark.tags.isNotEmpty() && showTags) {
            for (i in 0 until bookmark.tags.size) {
              val tagId = bookmark.tags[i]
              val tag = tagMap[tagId]
              tag?.let { tagList.add(it.copy(name = CollectionNames.displayName(context, it))) }
            }
          }

          if (tagList.isEmpty()) {
            tags.visibility = View.GONE
          } else {
            tags.setTags(tagList)
            tags.visibility = View.VISIBLE
          }
        }
      }
    }
  }

  private fun bindHeader(holder: HeaderHolder, pos: Int) {
    val item = elements[pos]
    holder.title.text = item.text
    val trailingNumber = item.itemCount ?: item.page
    if (trailingNumber == 0) {
      holder.pageNumber.visibility = View.GONE
    } else {
      holder.pageNumber.visibility = View.VISIBLE
      holder.pageNumber.text = QuranUtils.getLocalizedNumber(trailingNumber)
    }
    holder.setChecked(isItemChecked(pos))
    holder.setEnabled(isEnabled(pos))
  }

  private fun bindCollectionHeader(holder: CollectionHeaderHolder, pos: Int) {
    val item = elements[pos]
    holder.title.text = item.text

    val itemCount = item.itemCount
    if (itemCount == null) {
      holder.pageNumber.visibility = View.GONE
    } else {
      holder.pageNumber.visibility = View.VISIBLE
      holder.pageNumber.text = QuranUtils.getLocalizedNumber(itemCount)
    }

    if (item.isCollapsible) {
      holder.expandIcon.visibility = View.VISIBLE
      holder.expandIcon.setImageResource(
        if (item.isCollapsed) R.drawable.ic_expand_more_24 else R.drawable.ic_expand_less_24
      )
      holder.expandIcon.contentDescription = context.getString(
        if (item.isCollapsed) R.string.expand_collection else R.string.collapse_collection
      )
    } else {
      holder.expandIcon.visibility = View.INVISIBLE
    }

    holder.openIcon.visibility = if (item.tagId == null) View.GONE else View.VISIBLE
    holder.setChecked(isItemChecked(pos))
    holder.setEnabled(isEnabled(pos))
  }

  private fun bindHighlightColor(holder: HighlightColorHolder, pos: Int) {
    val item = elements[pos]
    holder.title.text = item.text

    val count = item.itemCount ?: 0
    holder.pageNumber.visibility = View.VISIBLE
    holder.pageNumber.text = QuranUtils.getLocalizedNumber(count)

    item.imageFilterColorResource?.let { colorResource ->
      holder.swatch.setColorFilter(
        ContextCompat.getColor(context, colorResource), PorterDuff.Mode.SRC_ATOP
      )
    }

    // a color with nothing in it stays in the list, but dims rather than shouting
    val alpha = if (count == 0) emptyColorAlpha else 1f
    holder.swatch.alpha = alpha
    holder.title.alpha = alpha
    holder.pageNumber.alpha = alpha

    holder.setChecked(isItemChecked(pos))
    holder.setEnabled(isEnabled(pos))
  }

  private fun isEnabled(position: Int): Boolean {
    val selected = elements[position]
    return !isEditable ||                     // anything in surahs or juzs
        selected.isBookmark ||                // actual bookmarks
        selected.isReadingBookmark ||         // the non-editable reading bookmark shortcut
        selected.rowType == QuranRow.NONE ||  // the actual "current page"
        selected.isHighlightColor ||          // the five highlight colors
        selected.isHighlightedAyah ||         // a highlighted ayah inside a color's screen
        selected.isBookmarkHeader             // tags
  }

  open inner class HeaderHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    val view: View = itemView
    val title: TextView = itemView.findViewById(R.id.title)
    val pageNumber: TextView = itemView.findViewById(R.id.pageNumber)

    fun setEnabled(enabled: Boolean) {
      view.isEnabled = true
      itemView.setOnClickListener(
        if (enabled) this@QuranListAdapter else null
      )
      itemView.setOnLongClickListener(
        if (isEditable && enabled) this@QuranListAdapter else null
      )
    }

    fun setChecked(checked: Boolean) {
      view.isActivated = checked
    }
  }

  private inner class ViewHolder(itemView: View) : HeaderHolder(itemView) {
    val metadata: TextView = itemView.findViewById(R.id.metadata)
    val number: TextView = itemView.findViewById(R.id.suraNumber)
    val image: ImageView = itemView.findViewById(R.id.rowIcon)
    val tags: TagsViewGroup = itemView.findViewById(R.id.tags)
    val date: TextView? = itemView.findViewById(R.id.show_date)
  }

  private inner class CollectionHeaderHolder(itemView: View) : HeaderHolder(itemView) {
    val expandIcon: ImageView = itemView.findViewById(R.id.expandIcon)
    val openIcon: ImageView = itemView.findViewById<ImageView>(R.id.openIcon).also { icon ->
      icon.setOnClickListener {
        val position = bindingAdapterPosition
        if (position != RecyclerView.NO_POSITION) {
          touchListener?.onOpenClicked(elements[position], position)
        }
      }
    }
  }

  private inner class HighlightColorHolder(itemView: View) : HeaderHolder(itemView) {
    val swatch: ImageView = itemView.findViewById(R.id.rowIcon)
  }

  interface QuranTouchListener {
    fun onClick(row: QuranRow, position: Int)
    fun onLongClick(row: QuranRow, position: Int): Boolean
    fun onOpenClicked(row: QuranRow, position: Int) = Unit
  }

  private companion object {
    private const val VIEW_TYPE_HEADER = 0
    private const val VIEW_TYPE_ROW = 1
    private const val VIEW_TYPE_COLLECTION_HEADER = 2
    private const val VIEW_TYPE_HIGHLIGHT_COLOR = 3
  }
}
